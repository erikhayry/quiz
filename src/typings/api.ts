export enum Category {
  Chemistry = "che",
  Economy = "eco",
  Peace = "pea",
  Phy = "phy",
  Medicine = "med",
}

type Translation = {
  en: string;
  se?: string;
  no?: string;
};

type ItemLink = {
  rel: string;
  href: string;
  action: string;
  types: string;
};

//TODO fix model https://app.swaggerhub.com/apis/NobelMedia/NobelMasterData/2.1#/default/get_nobelPrizes
type Laureate = {
  id: number;
  knownName?: Translation;
  orgName?: Translation;
  portion: string;
  sortOrder: string;
  motivation: Translation;
  links: ItemLink[];
};

export type NobelPrize = {
  awardYear: number;
  category: Category;
  categoryFullName: Translation;
  dateAwarded: string;
  prizeAmount: number;
  prizeAmountAdjusted: number;
  topMotivation: Translation;
  laureates: Laureate[];
};

type NobelPrizesMeta = {
  offset: number;
  limit: number;
  minimum: number;
  nobelPrizeYear: number;
  yearTo: number;
  nobelPrizeCategory: Category;
  count: number;
};

type Links = {
  first: string;
  prev: string;
  self: string;
  next: string;
  last: string;
};

type ApiResponse = {
  nobelPrizes: NobelPrize[];
  meta: NobelPrizesMeta;
  links: Links;
};
