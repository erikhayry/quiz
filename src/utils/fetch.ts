const headers = {
  "Content-Type": "application/json",
};

export const fetchAsJSON = async (url: string): Promise<any | undefined> => {
  try {
    const res = await fetch(url, {
      method: "GET",
      headers,
    });
    if (res.status !== 200) {
      console.error(`server responded with ${res.status} for ${url}`);
      return undefined;
    }
    const json = await res.json();

    return json || undefined;
  } catch (error) {
    console.error(error);

    return undefined;
  }
};
