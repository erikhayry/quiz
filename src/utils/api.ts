import { fetchAsJSON } from "./fetch";

const NOBEL_PRIZE_API_PATH = "https://api.nobelprize.org/2.1/nobelPrizes";

async function getNobelPrizes(): Promise<NobelPrize[]> {
  const res = await fetchAsJSON(
    `${NOBEL_PRIZE_API_PATH}?nobelPrizeYear=2000&yearTo=2020&limit=20&nobelPrizeCategory=pea`
  );

  if (res) {
    return res.nobelPrizes as NobelPrize[];
  }

  return [];
}

//TODO handle names
function mapToWinner({ awardYear, laureates }: NobelPrize): Winner {
  return {
    name: laureates[0].knownName?.en || laureates[0].orgName?.en || "UNKOWN",
    year: awardYear,
  };
}

export async function getWinners(): Promise<Winner[]> {
  const nobelPrizes = await getNobelPrizes();

  return nobelPrizes.map(mapToWinner);
}
