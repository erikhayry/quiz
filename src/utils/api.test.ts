import { getWinners } from "./api";
import { fetchAsJSON } from "./fetch";
import mockNobelPrizesResponse from "./mocks/nobelPrizes.json";

jest.mock("./fetch", () => ({
  fetchAsJSON: jest.fn(),
}));

describe("api", () => {
  describe("getWinners", () => {
    it("handles failed request", async () => {
      (fetchAsJSON as jest.Mock).mockReturnValue(undefined);

      const winners = await getWinners();
      expect(winners.length).toEqual(0);
    });

    it("returns winners", async () => {
      (fetchAsJSON as jest.Mock).mockReturnValue(mockNobelPrizesResponse);

      const winners = await getWinners();
      expect(winners).toMatchSnapshot();
    });
  });
});
