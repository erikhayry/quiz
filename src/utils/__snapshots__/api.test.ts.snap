// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api getWinners returns winners 1`] = `
Array [
  Object {
    "name": "Kim Dae-jung",
    "year": "2000",
  },
  Object {
    "name": "United Nations",
    "year": "2001",
  },
  Object {
    "name": "Jimmy Carter",
    "year": "2002",
  },
  Object {
    "name": "Shirin Ebadi",
    "year": "2003",
  },
  Object {
    "name": "Wangari Maathai",
    "year": "2004",
  },
  Object {
    "name": "International Atomic Energy Agency",
    "year": "2005",
  },
  Object {
    "name": "Muhammad Yunus",
    "year": "2006",
  },
  Object {
    "name": "Intergovernmental Panel on Climate Change",
    "year": "2007",
  },
  Object {
    "name": "Martti Ahtisaari",
    "year": "2008",
  },
  Object {
    "name": "Barack H. Obama",
    "year": "2009",
  },
  Object {
    "name": "Liu Xiaobo",
    "year": "2010",
  },
  Object {
    "name": "Ellen Johnson Sirleaf",
    "year": "2011",
  },
  Object {
    "name": "European Union (EU)",
    "year": "2012",
  },
  Object {
    "name": "Organisation for the Prohibition of Chemical Weapons",
    "year": "2013",
  },
  Object {
    "name": "Kailash Satyarthi",
    "year": "2014",
  },
  Object {
    "name": "National Dialogue Quartet ",
    "year": "2015",
  },
  Object {
    "name": "Juan Manuel Santos",
    "year": "2016",
  },
  Object {
    "name": "International Campaign to Abolish Nuclear Weapons (ICAN)",
    "year": "2017",
  },
  Object {
    "name": "Denis Mukwege",
    "year": "2018",
  },
  Object {
    "name": "Abiy Ahmed Ali",
    "year": "2019",
  },
]
`;
